File: src/main/java/com/gitlab/daring/image/event/ValueEvent.java
package com.gitlab.daring.image.event;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

public class ValueEvent<T> implements Consumer<T> {

    public boolean enabled = true;
    
    private final List<Consumer<T>> listeners = new CopyOnWriteArrayList<>();

    public void addListener(Consumer<T> l) {
        if (!listeners.contains(l))
            listeners.add(l);
    }

    public void removeListener(Consumer<T> l) {
        listeners.remove(l);
    }

    public void removeListeners() {
        listeners.clear();
    }

    public void fire(T value) {
        if (!enabled) return;
        listeners.forEach(c -> c.accept(value));
    }

    @Override
    public void accept(T v) {
        fire(v);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/event/ValueEvent.kt
package com.gitlab.daring.image.event

import java.util.concurrent.CopyOnWriteArrayList

open class ValueEvent<T> {

    var enabled = true

    private val listeners = CopyOnWriteArrayList<Listener<T>>()

    fun addListener(l: Listener<T>) {
        if (!listeners.contains(l))
            listeners.add(l)
    }

    fun removeListener(l: Listener<T>) {
        listeners.remove(l)
    }

    fun removeListeners() {
        listeners.clear()
    }

    fun fire(value: T?) {
        if (!enabled) return
        listeners.forEach { it.invoke(value) }
    }

}

typealias Listener<T> = (T?) -> Unit